editor/src/clj/editor/atlas.clj

```diff
@@ -1186,8 +1191,67 @@
       (scene-tools/scale-factor camera viewport (Vector3d. (first pivot-pos) (second pivot-pos) 0.0)))))
 
+(defn- image-path?
+  [path]
+  (some (partial str/ends-with? path) image/exts))
+
+(defn- get-image-resource-from-file
+  [workspace ^File file]
+  (when-let [path (workspace/as-proj-path workspace (.getAbsolutePath file))]
+    (when (image-path? path)
+      (workspace/resolve-workspace-resource workspace path))))
+
+(defn- get-image-resources-from-dragboard
+  [^Dragboard dragboard workspace]
+  (->> (.getFiles dragboard)
+       (keep (partial get-image-resource-from-file workspace))
+       (sort-by resource/path)))
+
+(defn- image-resources->image-msgs
+  [image-resources]
+  (mapv (partial hash-map :image) image-resources))
+
+(defn- create-dropped-images!
+  [parent image-resources op-seq]
+  (g/tx-nodes-added
+    (g/transact
+      (concat
+        (g/operation-sequence op-seq)
+        (g/operation-label "Drop images")
+        (condp g/node-instance? parent
+          AtlasNode (let [existing-image-resources (set (g/node-value parent :image-resources))
+                          new-image? (complement existing-image-resources)]
+                      (->> (filter new-image? image-resources)
+                           (image-resources->image-msgs)
+                           (make-image-nodes-in-atlas parent)))
+          AtlasAnimation (->> (image-resources->image-msgs image-resources)
+                              (make-image-nodes-in-animation parent)))))))
+
+(defn- parent-animation-or-atlas
+  [selection]
+  (or (first (handler/adapt-every selection AtlasAnimation))
+      (first (map #(core/scope-of-type % AtlasAnimation) selection))
+      (first (map #(core/scope-of-type % AtlasNode) selection))
+      (first (handler/adapt-every selection AtlasNode))))
+
 (defn handle-input [self action selection-data]
   (case (:type action)
+    :drag-dropped (let [dragboard ^Dragboard (:dragboard action)]
+                    (when (.hasFiles dragboard)
+                      (let [app-view (g/node-value self :app-view)
+                            selection (g/node-value app-view :selected-node-ids)
+                            parent (parent-animation-or-atlas selection)
+                            project (project/get-project parent)
+                            workspace (project/workspace project)
+                            image-resources (get-image-resources-from-dragboard dragboard workspace)
+                            op-seq (gensym)
+                            image-nodes (create-dropped-images! parent image-resources op-seq)
+                            drag-event ^DragEvent (:event action)]
+                        (when (seq image-nodes)
+                          (select! app-view image-nodes op-seq)
+                          (ui/user-data! (ui/main-scene) ::ui/refresh-requested? true)
+                          (.setDropCompleted drag-event true))
+                        nil)))
```

Comment:
I think you should be able to get what you need here from the ui context of the `:target` of the `action`, to avoid connecting the `app-view` to each `ToolController`.
```clj
(let [image-view (:target action) ; JavaFX ImageView that was the drop target.
      ui-context (first (ui/node-contexts image-view false))
      {:keys [app-view project selection workspace]} (:env ui-context)]
  ...)
```